# erro

Uma [Ferramenta Sumor Cloud](https://sumor.cloud).  
[Documentação Adicional](https://sumor.cloud/error)
Uma biblioteca de tratamento de erros com suporte a vários idiomas, código de erro predefinido e passagem de dados.

[![CI](https://github.com/sumor-cloud/error/actions/workflows/ci.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ci.yml)
[![Teste](https://github.com/sumor-cloud/error/actions/workflows/ut.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ut.yml)
[![Cobertura](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml)
[![Auditoria](https://github.com/sumor-cloud/error/actions/workflows/audit.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/audit.yml)

## Instalação

```bash
npm i @sumor/error --save
```

## Pré-requisitos

### Versão do Node.JS

Requer Node.JS na versão 18.x ou superior

### Requerir módulo do Node.JS ES

Como este pacote é escrito em módulo ES,
por favor altere o seguinte código em seu arquivo `package.json`:

```json
{
  "type": "module"
}
```

## Uso

### Definir Erro

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  code: {
    USUÁRIO_NÃO_ENCONTRADO: 'Usuário não encontrado',
    USUÁRIO_EXISTENTE: 'Usuário {name} existente'
  }
})

throw new MeuErro('USUÁRIO_NÃO_ENCONTRADO')
// saída: Error: Usuário não encontrado

throw new MeuErro('USUÁRIO_EXISTENTE', { name: 'Alice' })
// saída: Error: Usuário Alice existente
```

### Multilinguagem

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  language: 'en', // idioma padrão
  code: {
    USUÁRIO_NÃO_ENCONTRADO: 'Usuário não encontrado',
    USUÁRIO_EXISTENTE: 'Usuário {name} existente'
  },
  i18n: {
    zh: {
      USUÁRIO_NÃO_ENCONTRADO: 'Usuário não encontrado',
      USUÁRIO_EXISTENTE: 'Usuário {name} existente'
    }
  }
})

const erro = new MeuErro('USUÁRIO_EXISTENTE', { name: 'Alice' })
erro.language = 'en' // alterar idioma do erro
console.log(erro)
// saída: Error: Usuário Alice existente

erro.language = 'zh' // alterar idioma do erro
console.log(erro)
// saída: Error: Usuário Alice existente
```

### Converter Erro para JSON

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  code: {
    USUÁRIO_NÃO_ENCONTRADO: 'Usuário não encontrado',
    USUÁRIO_EXISTENTE: 'Usuário {name} existente'
  }
})

const erro = new MeuErro('USUÁRIO_EXISTENTE', { name: 'Alice' })
console.log(erro.json())
// saída: {"code":"USUÁRIO_EXISTENTE","message":"Usuário Alice existente"}
```

### Erro Subjacente

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  code: {
    VERIFICAÇÃO_CAMPO_FALHOU: 'Falha na verificação do campo',
    CAMPO_NÃO_PODE_VAZIO: 'Campo {name} não pode ser vazio',
    CAMPO_MUITO_LONGO: 'Campo {name} é muito longo'
  },
  i18n: {
    zh: {
      VERIFICAÇÃO_CAMPO_FALHOU: 'Falha na verificação do campo',
      CAMPO_NÃO_PODE_VAZIO: 'Campo {name} não pode ser vazio',
      CAMPO_MUITO_LONGO: 'Campo {name} é muito longo'
    }
  }
})

const erro = new MeuErro('VERIFICAÇÃO_CAMPO_FALHOU', {}, [
  new MeuErro('CAMPO_NÃO_PODE_VAZIO', { name: 'nome de usuário' }),
  new MeuErro('CAMPO_MUITO_LONGO', { name: 'senha' })
])

console.log(erro.json())
/* 
saída: 
{
  "code":"VERIFICAÇÃO_CAMPO_FALHOU",
  "message":"Falha na verificação do campo",
  "errors":[
    {
      "code":"CAMPO_NÃO_PODE_VAZIO",
      "message":"Campo nome de usuário não pode ser vazio"
    },{
      "code":"CAMPO_MUITO_LONGO",
      "message":"Campo senha é muito longo"
    }
  ]
}
*/

erro.language = 'zh'
console.log(erro.json())
/*
saída:
{
  "code":"VERIFICAÇÃO_CAMPO_FALHOU",
  "message":"Falha na verificação do campo",
  "errors":[
    {
      "code":"CAMPO_NÃO_PODE_VAZIO",
      "message":"Campo nome de usuário não pode ser vazio"
    },{
      "code":"CAMPO_MUITO_LONGO",
      "message":"Campo senha é muito longo"
    }
  ]
}
*/
```

### Combinar Erro Padrão

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  code: {
    VERIFICAÇÃO_CAMPO_FALHOU: 'Falha na verificação do campo',
    CAMPO_NÃO_PODE_VAZIO: 'Campo {name} não pode ser vazio',
    CAMPO_MUITO_LONGO: 'Campo {name} é muito longo'
  }
})

const erro = new MeuErro('VERIFICAÇÃO_CAMPO_FALHOU', {}, [
  new MeuErro('CAMPO_NÃO_PODE_VAZIO', { name: 'nome de usuário' }),
  new MeuErro('CAMPO_MUITO_LONGO', { name: 'senha' }),
  new Error('Erro Desconhecido')
])

console.log(erro.json())
/*
saída:
{
  "code":"VERIFICAÇÃO_CAMPO_FALHOU",
  "message":"Falha na verificação do campo",
  "errors":[
    {
      "code":"CAMPO_NÃO_PODE_VAZIO",
      "message":"Campo nome de usuário não pode ser vazio"
    },{
      "code":"CAMPO_MUITO_LONGO",
      "message":"Campo senha é muito longo"
    },{
      "code":"ERRO_DESCONHECIDO",
      "message":"Erro Desconhecido"
    }
  ]
}
 */
```
